### ISSUE TITLE
Community App Performance : Server Calls

### ISSUE THREAD

**Issue by sushilshinde** [topcoder-platform/community-app#1138]
The goal of this report is to identify non-essential components that are in between the user and first user interaction.

**Browser**: Google Canary
**URL**: https://beta-community-app.topcoder.com/challenges
**Audit Tool:** Chrome Lighthouse
**Lighthouse Audit Profile**: Desktop, Simulated Fast 3G, 4x slowdown CPU throttling

It took more than 18 secs to become page interactive for the logged-in user.

![screen shot 2018-07-31 at 1 14 01 pm](https://user-images.githubusercontent.com/683231/43445290-a7cb0284-94c3-11e8-84e9-0294c355d228.png)

 
![screen shot 2018-07-31 at 1 12 08 pm](https://user-images.githubusercontent.com/683231/43445185-6026de9e-94c3-11e8-8731-867d0ad6e30a.png)

### 0 - 1 secs
**700ms** : DNS query, TCP connection, TLS connection, TCP congestion algo stuff. 

**Fix :** Lot can't be done here but connection reuse with http2 binary/multichannel/stream leverage, using prefetch, dns-prefetch, preload attributes. This will help browser resolve DNS and complete TCP, TLS handshakes eagrly. Most assets required for this page can be fetched like this.
	
**750ms**: 

CSS requests start, these files are huge and lot of code is unused.  

Members profile image call takes 153 ms can be deferred. Loading indicator js is a bootstrap JS and calls main-XXX.js after some time. 

Font calls: These are long calls will be taken care by PWA caching, nothing to do here. Font files are hardly changed.

**Spinner still on the screen**

**Fix**

1. Needs build treeshaking for JS and CSS files 
I have seen serious unused bytes of CSS files more than 82% of the payload isn't being used.

![screen shot 2018-07-31 at 2 03 18 pm](https://user-images.githubusercontent.com/683231/43448053-83ae34e6-94ca-11e8-891a-2bee709af988.png)

More than 50% of JS payload isn't being used

![screen shot 2018-07-31 at 2 07 24 pm](https://user-images.githubusercontent.com/683231/43448269-23305db4-94cb-11e8-901c-9accb9e24050.png)

If you want to see what's the cost check out this report which compiles [cost of imports](https://bundlephobia.com/scan-results?packages=atob@2.1.0,babel-register@6.26.0,babel-runtime@6.26.0,body-parser@1.18.3,browser-cookies@1.1.0,classnames@2.2.5,config@1.30.0,cookie-parser@1.4.3,d3@3.5.17,draft-js@0.10.4,draft-js-export-html@1.2.0,draft-js-markdown-shortcuts-plugin@0.3.0,draft-js-plugins-editor@2.0.3,express@4.16.3,filestack-react@1.3.8,helmet@3.12.1,html-to-text@3.3.0,i18n-iso-countries@3.6.3,immutable@3.8.2,isomorphic-fetch@2.2.1,le_node@1.7.0,lodash@4.17.10,markdown-it@8.4.0,moment@2.22.2,moment-duration-format@2.2.2,moment-timezone@0.5.17,money@0.2.0,morgan@1.9.0,node-forge@0.7.5,nuka-carousel@3.0.1,prop-types@15.6.2,qs@6.5.1,raf@3.4.0,rc-tooltip@3.4.9,react@16.4.1,react-color@2.13.8,react-css-super-themr@2.2.0,react-dates@12.6.0,react-dom@16.4.1,react-helmet@5.2.0,react-player@0.24.1,react-redux@5.0.7,react-redux-toastr@7.2.6,react-router-dom@4.3.1,react-select@1.1.0,react-slick@0.15.4,react-stickynode@1.4.1,react-svg@4.0.17,react-waypoint@7.1.0,redux@3.7.2,redux-actions@2.4.0,redux-promise@0.6.0,request-ip@2.0.2,require-context@1.1.0,serialize-javascript@1.5.0,serve-favicon@2.5.0,sharp@0.20.0,shortid@2.2.8,showdown@1.8.6,slick-carousel@1.8.1,supertest@3.0.0,topcoder-react-lib@0.5.0-2,topcoder-react-ui-kit@0.4.3,topcoder-react-utils@0.6.7,turndown@4.0.1,url-parse@1.4.1,uuid@3.0.1,valid-url@1.0.9,xml2json@0.11.0
) of bundles use in community app.

2. Use `import` wisely [here is the VS code plugin to measure it at code level](https://marketplace.visualstudio.com/items?itemName=wix.vscode-import-cost). It can be done at code level but seems like big work. Build time treeshaking should solve this problem.
3. Use JS treeshaking (topcoder react util)
4. Use CSS treeshaking (topcoder react util)
5. This can be a preload/prefetch call.

### 1 - 2 secs	
This is mostly the idle phase, nothing much is happening except waterfall calls to vendor libraries. 

![screen shot 2018-07-31 at 7 30 57 pm](https://user-images.githubusercontent.com/683231/43464318-6f534c06-94f8-11e8-9645-b294c2f7fa29.png)

**1000ms:** Google analytics JS call, Zendesk. For the first load, they can be deferred and Cache these with stale strategies in PWA
			
**2100ms:**  Challenge listing js call, where dependency main js and polyfill is already finished loading at 1200ms, there is a huge gap. 
	
**Spinner still on the screen**	

**Fix**
These calls can be prefetched, we aren't using the Http2  advantage that CDN is offering. Also, these scripts can be preloaded/prefetch. This phase is preceded by huge huge scripting (yellow region)
	
### 2 - 18 secs
This is mostly idle phase of the whole report because the way business data API calls are made.

**2400ms:** First API call
**2700ms:** Spinner stops and empty app shell is displayed 
		
Rest of this phase performs a series of API calls in a waterfall style. 

**Fix**
These calls can be reduced or prefetched as these are fixed URL calls no param values are changed every time they are called. As I see these are http2 calls, they can be prefetched in parallel, the browser will same connection and it will be fast. Here on it take Approx 15000ms to become page interactive as per the report.

Report: Use the above lighthouse profile to generate, it will differ a little but the pattern of problems will be the same.
